### PR TITLE
Merge Sidebars take 2

### PR DESCRIPTION
**Pull request by ashleemboyer**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1202022787106807/1202132849199133/f) 🎟️

## What/Why

- Renames `BackToLinkProps` `url` to `href` so it's a little more consistent with other components that render links.
- Updates `Sidebar` to accept either `children` or `menuItems`, and takes preference over `children`. This makes allows customization of the portion of `Sidebar` that does not need to be consistent across all `Sidebar`s.
- Renames `TutorialsSidebar`'s `backToLink` prop to `backToLinkProps` for consistent naming that's more traceable to the prop's intended use.
- Updates `TutorialsSidebar` to render `Sidebar` and pass it `children`. No updates where made to the `children` that are rendered.
- Adds `visuallyHideTitle` prop to `Sidebar` and the `g-screen-reader-only` global CSS helper class. These make it possible to visually hide the title heading of the `Sidebar`.
- Updates `TutorialSidebarProps` to reference `SidebarProps` for the props that are just sent straight to `Sidebar` without modification. This makes the props more traceable to their intended use.

## How

Nothing unique to add here that I can think of, but happy to answer any questions that come up!

### Product landing pages

The sidebars on the following pages should still render a Sidebar title, menu items, and a Resources section:

- [ ] [/vault](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/vault)
- [ ] [/waypoint](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint)

### Docs sidebars

The sidebars on the following pages should still render a back to link, filter input, Sidebar title, menu items, and a Resources section:

- [ ] [/vault/docs](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/vault/docs)
- [ ] [/vault/api-docs](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/vault/api-docs)
- [ ] [/waypoint/docs](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint/docs) 
- [ ] [/waypoint/commands](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint/commands)
- [ ] [/waypoint/plugins](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint/plugins)

### CollectionView Sidebars (`/{productSlug}/tutorials/{collectionSlug}`)

The sidebars on the following pages should still render a back to link, Sidebar title, an Overview menu item, sections of collection categories with menu items, and a Resources section:

- [ ] [/vault/tutorials/adp](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/vault/tutorials/adp)
- [ ] [/waypoint/tutorials/application-deployment](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint/tutorials/application-deployment)
- [ ] Check additional collection views, these are just examples

### TutorialView Sidebars (`/{productSlug}/tutorials/{collectionSlug}/{tutorialSlug}`)

The sidebars on the following pages should still render visually hidden Sidebar title, menu items, and a Resources section:

- [ ] [/vault/tutorials/adp/transform](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/vault/tutorials/adp/transform)
- [ ] [/waypoint/tutorials/get-started-docker/get-started-intro](https://dev-portal-git-ambmerge-sidebars-v2-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-intro)
- [ ] Check additional tutorial views, these are just examples

## Anything else?

This PR stemmed from #329 and the conversation that happened after it.

### CODE REVIEW

File: src/layouts/sidebar-sidecar/index.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ const SidebarSidecarLayout: FC<SidebarSidecarLayoutProps> = ({
   sidebarProps,
   sidebarSlot,
   sidecarSlot,
-}) => {
+}: SidebarSidecarLayoutProps) => {
```
Comment (ashleemboyer) [inline src/layouts/sidebar-sidecar/index.tsx:20]: These changes are a `typedef` fix.

File: src/components/sidebar/types.ts [hunk 2]
```diff
@@ -44,4 +40,34 @@ export interface SidebarProps {
    * Text to be shown as the title of the sidebar.
    */
   title: string
+
+  /**
+   * @TODO - write description
```
Comment (zchsh) [inline src/components/sidebar/types.ts:45]: Non-blocking: we could drop in the description from `TutorialsSidebar`:

```suggestion
   * Optional. If true, the title of the sidebar will be visually hidden.
```

File: src/pages/style.css [hunk 3]
```diff
@@ -129,6 +129,26 @@ Mobile & tablet visibility helpers, specific to dev-dot.
...
+***
+* Screen reader only global helper
+***
+* Visually hides content so it is only available to screen readers. Sourced
+* Scott O'Hara's Inclusively Hidden article.
+*
+* https://www.scottohara.me/blog/2017/04/14/inclusively-hidden.html
+***
+*/
+.g-screen-reader-only:not(:focus):not(:active) {
```
Comment (zchsh) [inline src/pages/style.css:142]: Non-blocking question: going forward do you think we should aim to use this `className` rather than `VisuallyHidden`?

**Review by zchsh — APPROVED** [review]
✅ this looks good to me!

It feels like this new, merged `components/sidebar` component resolves the duplicated `nav` element, yet also introduces some conditionals around filtering and adding metadata (for example). I think this might be because significant parts of `components/sidebar` are docs specific 🤔 Might also be related to the optional `menuItems` prop...

Very likely outside the scope of this PR, but I wonder if there's a way we could more cleanly split the components & functionality needed for `docs`, `tutorials`, and more general sidebar content.

File: src/pages/style.css (hunk 3 above)
Comment (ashleemboyer) [reply src/pages/style.css:142]: Yep!

**Comment by ashleemboyer** [comment]
> It feels like this new, merged `components/sidebar` component resolves the duplicated `nav` element, yet also introduces some conditionals around filtering and adding metadata (for example). I think this might be because significant parts of `components/sidebar` are docs specific 🤔 Might also be related to the optional `menuItems` prop...

Yep, it has everything to do with the now optional `menuItems` prop. The following comment in `src/components/sidebar/types.ts` captures the context here:

```js
/**
 * `SidebarContentProps` defines the properties that represent what content is
 * rendered inside of `Sidebar` below the consistent "header" section used in
 * all instances of `Sidebar`.
 *
 * This approach allows us to require either (not both) `children` and
 * `menuItems` since providing both of these props is not a case that this
 * component handles.
 *
 * TODO: we may decide to remove `menuItems` from instances of `Sidebar` that
 * used that prop before the `children` prop was introduced. For now, we will
 * support either prop.
 */
```

> Very likely outside the scope of this PR, but I wonder if there's a way we could more cleanly split the components & functionality needed for `docs`, `tutorials`, and more general sidebar content.

There's a generalized `TODO` for this, see previous. 😊

> EDIT: random non-blocking note: on the product landing pages, I'm not seeing `back to link`s as described in the `Testing` section, but I think this is expected, might just be a slip in the testing section text 👍

Oh right, those were changed with #331, hadn't updated that part of the test steps I pulled from #329. Great catch! ✨

File: src/components/sidebar/types.ts (hunk 2 above)
Comment (ashleemboyer) [reply src/components/sidebar/types.ts]: Updated in https://github.com/hashicorp/dev-portal/pull/335/commits/fc75a3ec2ff11a7c1d458964440d6c940207074f!

**Comment by ashleemboyer** [comment]
@zchsh @BRKalow I had to make a couple of updates for the `overviewItemHref` prop, so this could use a re-review:

- commit from merging `main` into this branch: https://github.com/hashicorp/dev-portal/pull/335/commits/640dd38a0cc3d50253c9c205baec08dadbd568b1
- commit handling which component wraps nav item content with an `<li>`, needed so that the new Overview item does not get wrapped in `<li>`: https://github.com/hashicorp/dev-portal/pull/335/commits/2502475872996264380ea8f242f9cfbe8a12ccf4

**Review by BRKalow — APPROVED** [review]
Looks great! I think this is a fantastic iterative step for our sidebar architecture 🌈  I'm optimistic we can continue to lean into the composition-based approach as we continue to layer on additional sidebar features ❤️ 

Maybe good to have @zchsh and or @kendallstrautman also put eyes on this before merge?

**Review by zchsh — APPROVED** [review]
✅ Looks great to me! +1 to what Bryce said 💖

**Review by kendallstrautman — APPROVED** [review]

**Pull request merged**